src/backends/backend.cpp

```diff
@@ -0,0 +1,616 @@
+#include "library/normalize.h"
+#include "library/let.h"
+#include "library/trace.h"
+#include "used_names.h"
+#include "util/fresh_name.h"
+#include "util/name.h"
+#include "util/name_set.h"
+
+namespace lean {
+
+bool is_erasible(expr const & e);
+
+// Probably should use buffer everywhere. TODO
+std::vector<name> to_vector(buffer<name> & names) {
+    std::vector<name> rs;
+    for (auto n : names) {
+        rs.push_back(n);
+    }
+    return rs;
+}
+
+backend_exception not_supported(char const * msg) {
+    return backend_exception(msg);
+}
+
+backend_exception internal_error(char const * msg) {
+    return backend_exception(msg);
+}
+
+backend::backend(environment const & env, optional<std::string> main_fn)
+    : m_env(env), m_tc(m_env), m_debug_tracing(true) {
+    auto main_name = name(main_fn.value());
+    auto main = env.get(main_name);
+
+    // Compute the live set of names.
+    used_defs used_names(env);
+    used_names.names_in_decl(main);
+    auto decls_to_compile = std::vector<declaration>();
+    used_names.m_used_names.for_each([&] (name const &n) {
+        decls_to_compile.push_back(env.get(n));
+    });
+
+    // Perform the expr -> expr optimizations.
+
+    bool error_occured = false;
+
+    // Compile the decls into simple_exprs and procs.
+    for (auto decl : decls_to_compile) {
+        try {
+            this->compile_decl(decl);
+        } catch (const backend_exception & e) {
+            std::cout << e.what() << std::endl;
+            error_occured = true;
+        } catch (backend_exception * e) {
+            std::cout << e->what() << std::endl;
+            error_occured = true;
+        }
+    }
+
+    // if (error_occured) {
+    //     exit(1);
+    // }
+
+    // Perform the simple_expr -> simple_expr optimizations.
+}
+
+void backend::compile_decl(declaration const & d) {
+    if (d.is_axiom() && is_extern(this->m_env, d.get_name())) {
+        std::cout << "extern: " << std::endl;
+        std::cout << d.get_name() << std::endl;
+    } else if (d.is_definition()) {
+        auto ty = d.get_type();
+        auto value = d.get_value();
+
+        std::vector<name> args;
+        auto se = this->compile_body(args, value);
+        auto p = proc(d.get_name(), args, se);
+        this->add_proc(p);
+    } else if (d.is_constant_assumption()) {
+        if (auto n = inductive::is_intro_rule(this->m_env, d.get_name())) {
+            compile_intro_rule(d);
+        } else if (is_extern(this->m_env, d.get_name())) {
+            std::cout << "extern: " << d.get_name() << std::endl;
+        }
+    } else if (d.is_theorem()) {
+        std::stringstream ss;
+        ss << "internal_error: should not be trying to compile a theorem (";
+        ss << d.get_name();
+        ss << ")";
+        throw new backend_exception(ss.str().c_str());
+    } else {
+        std::stringstream ss;
+        ss << "unhandled declaration: ";
+        ss << d.get_name();
+        throw new backend_exception(ss.str().c_str());
```

Comment:
I think you should throw by value and catch by reference. First of all you are leaking memory, and more importantly your catch look uglier because it deals with two catch doing essentially the same - only one work on backend_exception*, one work on backend_exception. see http://www.informit.com/articles/article.aspx?p=373339